processing/src/main/java/org/apache/druid/segment/join/Joinables.java

```diff
@@ -114,4 +118,37 @@ public static boolean isPrefixedBy(final String columnName, final String prefix)
       );
     }).collect(Collectors.toList());
   }
+
+  public static void checkPreJoinableClausesForDuplicatesAndShadowing(
+      final List<PreJoinableClause> preJoinableClauses
+  )
+  {
+    List<String> prefixes = new ArrayList<>();
+    for (PreJoinableClause clause : preJoinableClauses) {
+      prefixes.add(clause.getPrefix());
+    }
+
+    checkPrefixesForDuplicatesAndShadowing(prefixes);
+  }
+
+  public static void checkPrefixesForDuplicatesAndShadowing(
+      final List<String> prefixes
+  )
+  {
+    for (int i = 0; i < prefixes.size(); i++) {
+      String prefix1 = prefixes.get(i);
+      for (int k = 0; k < prefixes.size(); k++) {
```

Comment:
If the list is sorted by size (largest -> smallest) we can change this loop to `k = i` instead of `k = 0`

Reply:
Added a sort and the `k = i` optimization